motorutil/steer.go

```diff
@@ -0,0 +1,255 @@
+		Err()
+	if s.err != nil {
+		return s
+	}
+	s.err = s.Right.
+		SetSpeedSetpoint(rightSpeed).
+		SetPositionSetpoint(rightCounts).
+		Err()
+	if s.err != nil {
+		return s
+	}
+
+	s.err = s.Left.Command("run-to-rel-pos").Err()
+	if s.err != nil {
+		return s
+	}
+	s.err = s.Right.Command("run-to-rel-pos").Err()
+	if s.err != nil {
+		s.Left.Command("stop").Err()
+	}
+	return s
+}
+
+// SteerDuration steers in the given turn for the given duration, d,  and at the
+// specified speed. The valid range of turn is -100 (hard left) to +100 (hard right).
+// If speed is negative, the turn will be made in reverse.
+//
+// See the ev3dev.SetSpeedSetpoint and ev3dev.SetTimeSetpoint documentation for speed
+// and duration behaviour.
+func (s *Steering) SteerDuration(speed, turn int, d time.Duration) *Steering {
+	if s.err != nil {
+		return s
+	}
+
+	if turn < -100 || 100 < turn {
+		s.err = directionError(turn)
+		return s
+	}
+	if d < 0 {
+		s.err = durationError(d)
+		return s
+	}
+
+	leftSpeed, _, rightSpeed, _ := motorRates(speed, turn, 0)
+
+	s.err = s.Left.
+		SetSpeedSetpoint(leftSpeed).
+		SetTimeSetpoint(d).
+		Err()
+	if s.err != nil {
+		return s
+	}
+	s.err = s.Right.
+		SetSpeedSetpoint(rightSpeed).
+		SetTimeSetpoint(d).
+		Err()
+	if s.err != nil {
+		return s
+	}
+
+	s.err = s.Left.Command("run-timed").Err()
+	if s.err != nil {
+		return s
+	}
+	s.err = s.Right.Command("run-timed").Err()
+	if s.err != nil {
+		s.Left.Command("stop").Err()
+	}
+	return s
+}
+
+func motorRates(speed, turn, counts int) (leftSpeed, leftCounts, rightSpeed, rightCounts int) {
+	switch {
+	case turn == 0:
+		leftSpeed = speed
+		rightSpeed = speed
+		leftCounts = counts
+		rightCounts = counts
+	case turn < 0:
+		rightSpeed = speed
+		rightCounts = counts
+		turn = (turn + 50) * 2
+		leftSpeed = (speed * turn) / 100
+		leftCounts = (rightCounts * turn) / 100
+	case turn > 0:
+		leftSpeed = speed
+		leftCounts = counts
+		turn = (50 - turn) * 2
+		rightSpeed = (speed * turn) / 100
+		rightCounts = (leftCounts * turn) / 100
+	}
+	return leftSpeed, leftCounts, rightSpeed, rightCounts
+}
+
+// Err returns the error state of the Steering and clears it.
+func (s *Steering) Err() error {
+	err := s.err
+	s.err = nil
+	return err
+}
+
+// Wait waits for the last steering operation to complete. A non-nil error will either
+// implement the Cause method, which may be used to determine the underlying cause, or
+// be an Errors holding errors that implement the Cause method.
+func (s *Steering) Wait() error {
+	if err := s.Err(); err != nil {
+		return err
+	}
+
+	var errors [2]error
+
+	var wg sync.WaitGroup
+	for i, motor := range []struct {
+		side   string
+		device *ev3dev.TachoMotor
+	}{
+		{side: "left", device: s.Left},
+		{side: "right", device: s.Right},
+	} {
+		i := i
+		side := motor.side
+		device := motor.device
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			stat, ok, err := ev3dev.Wait(s.Left, ev3dev.Running, 0, 0, false, s.Timeout)
```

Comment:
`s.Left` should be `device`.

Reply:
Thanks. That clears up that bit of confusion.

Fixed.